Refactor clean up logger

jdabapo opened the pull request:
Refactor typing on playByPlay logger to reduce amount of code reproduction. however, caused some initial bugs, still working on this. wanted to keep this branch open just to see if you were open to this, as this is somewhat big? And doing this would imply you would want more changes to typing for the logger, which could cause issues if not done correctly

dumbmatter commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -5,7 +5,11 @@ import getInjuryRate from "./getInjuryRate";
 import { maxBy, range } from "../../../common/utils";
-import PlayByPlayLogger from "./PlayByPlayLogger";
+import PlayByPlayLogger, {
+	type blockType,
+	type fgaType,
+	type fgMissType,
```
Types should start with a capital letter, that's just the normal convention everyone uses.

dumbmatter commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -1953,25 +1957,17 @@ class GameSim extends GameSimBase {
 		let probMissAndFoul;
 		let type: ShotType;
-
+		let fgaLogType: fgaType = "fgaTipIn";
```
For all the lines like this... ideally you don't have to set a default value here. If you have code like this:

```
let x: "a" | "b";
if (condition) {
  x = "a";
} else {
  x = "b";
}
```

then TypeScript is smart enough to figure out that `x` has been assigned some valid value. If some branch is missing (like there was no `else` above) then TypeScript will give an error when you use the variable, because it might not be initialized.

If there is supposed to be some non-initialized case, then make it `fgaType | undefined`, and explicitly handle the `undefined` case when `fgaLogType` is used.

What you have now is worse than either of the scenarios I just described, because you might accidentally not be setting the variable in the if/else logic, and then not realize the default is accidentally falling through.

If it is actually intended for the default to fall through in some case... that's also confusing and would be better off designed differently.

Apologies if my code is making you do weird stuff here, I didn't look too closely!

jdabapo replied:
Ah yes, I originally had what you had in the code block, but I remember I had to change it since I missed a typing and it was falling through as undefined (which should not have been the case)! I updated the types so it should be okay to do this now.

dumbmatter commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -2033,47 +2015,46 @@ class GameSim extends GameSimBase {
+				pid: this.team[this.o].player[p].id,
+				clock: this.t,
+				desperation:
+					(fgaLogType as string) === "fgaTp" ||
+					(fgaLogType as string) === "fgaTpFake"
+						? rushed && forceThreePointer
+						: undefined,
+				pidPass:
+					(fgaLogType as string) === "fgaTipIn" && passer
```
Casting like `as string` should be avoided if possible since often it results in bugs. Is it really needed here?

dumbmatter commented on src/worker/core/GameSim.basketball/PlayByPlayLogger.ts:
```diff
@@ -136,49 +130,12 @@ type PlayByPlayEventInputNoScore =
 			t: TeamNum;
 			pid: number;
 			clock: number;
+			desperation?: boolean | undefined; // needed for tp/tpFake. undefined for others
+			pidPass?: number | undefined; // needed for tipin. undefined for others
```
Shorter code is nice, but not at the expense of type correctness. It's worth having 3 separate types here rather than combining it all into one. The whole point of TypeScript is that comments like "needed for tp/tpFake. undefined for others" should not be needed, instead TypeScript handles it for you automatically :)

dumbmatter reviewed (changes requested):
Less repetitive code would definitely be welcome, just need to make sure the TypeScript types don't get worse :)

Also if you hadn't noticed already... game sim code is probably the worst code in the game. I think two reasons... it's the oldest part of the codebase, and I didn't abstract it well enough when making other sports so often fixing things needs to be done 4 times! There's been a little improvement over the years, but it's still not great.

jdabapo commented:
Addressed comments, and made a decent amount of changes to the types... is a larger PR, and I didn't see any pre-existing tests for these changes. If needed I can try and write some tests for the logger and check if the functionality is all there?

jdabapo commented on src/worker/core/GameSim.basketball/PlayByPlayLogger.ts:
```diff
@@ -1,292 +1,100 @@
+		| "fgTp"
+		| "fgTpAndOne"
+		| "fgTipIn"
+		| "fgTipInAndOne"; // Wondering, why is there a missTpFake but no fgTpFake?
```
left this comment as a question, there was no pre existing fgaTPFake or fgTpFake

jdabapo replied:
Hm I originally included it in case there are some shenanigans with three pointers set to off, and a log needs to be included for it but that doesnt seem to be the case. Ill take this out!

jdabapo commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -2167,57 +2160,38 @@ class GameSim extends GameSimBase {
-			});
+			fgMissLogType = g.get("threePointers") ? "missTp" : "missTpFake";
+		} else {
+			throw new Error(
```
Probably should not crash the browser if log doesnt exist... maybe better to just do console.warn?

dumbmatter replied:
It's actually good because people will notice and it'll get fixed quick. Console messages will get ignored, but could indicate bugs.

For stuff like this I do `throw new Error("Should never happen");` so it's not wasting bytes on custom error messages that are likely never displayed. The error in the console (or in my error logger) will include the file and line number.

jdabapo replied:
okay sounds good, and quick question, is storage not very readily available for error logging? Is that a design constraint to keep in mind when writing new code/logs?

dumbmatter replied:
That's not about storage, it's about bundle size. Of course it's a very small difference, probably doesn't matter! But it's also about not having to spend mental energy on writing nice error messages for a case like "this is never going to happen, and maybe I'm just writing this to tell TypeScript that it's never going to happen".

Overall, I guess it's just personal preference and consistency.

jdabapo replied:
Interesting! I think its a bit different at work, where we write a ton of logs and description in logs. I think its also because this is primarily a solo dev project, as opposed to one where there's multiple teams working on multiple repositories. But good to know!

dumbmatter commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -2258,51 +2232,28 @@ class GameSim extends GameSimBase {
 		const p2 = this.playersOnCourt[this.d][pickPlayer(ratios)];
 		this.recordStat(this.d, p2, "blk");
-
+		let blockLogType: BlockType | undefined = undefined;
```
Nit, but you can leave out the ` = undefined` and it works the same

dumbmatter commented on src/worker/core/GameSim.basketball/PlayByPlayLogger.ts:
```diff
@@ -393,7 +200,8 @@ const scoringTypes = [
 	"fgMidRangeAndOne",
 	"ft",
-	"tp",
+	"fgTp",
+	"fgTpAndOne",
```
Switch back to tp. Ideally use TypeScript to guarantee there are no typos like this. (Also thanks for finding the missing tpAndOne - I wonder if that actually caused any problems...)

Even better - turn the array into a set, cause Set.has is slightly faster than Array.includes.

dumbmatter commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -2338,87 +2287,83 @@ class GameSim extends GameSimBase {
+		}
+		let fgMakeLogType:
+			| FgMakeType
+			| "fgPutBackAndOne"
+			| "fgPutBack"
+			| "fgAtRimAndOne"
+			| "fgAtRim";
```
I think this would be cleaner if those 4 extra ones were in `FgMakeType` and then you wouldn't have to manually type them out here (and not make a typo).

Could be like:

```
type FgMakeNormal = (what you have in FgMakeType now)
type FgMakeWithoutAst = "fgPutBackAndOne" | "fgPutBack";
type FgMakeWithDefender = "fgAtRimAndOne" | "fgAtRim";
type FgMakeType = FgMakeNormalType | FgMakeWithoutAst | FgMakeWithDefender;
```

And then use those when defining `PlayByPlayEventInputScore`.

That'd be just a little clearer and less error prone, I think.

dumbmatter commented:
Thanks, fantastic work to remove like 250 lines of code!

I did a few more little commits after merging, at least one is important and another is maybe semi important https://github.com/zengm-games/zengm/compare/362eee3e7e3c8a2f887b79bea48fac5a997d5cae...eeeda10d64408fe47128b8b6894ef7cb3189d044

jdabapo commented on src/worker/core/GameSim.basketball/index.ts:
```diff
@@ -2338,87 +2287,78 @@ class GameSim extends GameSimBase {
-
 			this.recordStat(this.o, p, "tpa");
 			this.recordStat(this.o, p, "tp");
+			fgMakeLogType = andOne ? "tp" : "tpAndOne";
```
Aw shoot I thought it was bug free my bad 🙃. Is there any roadmap to adding tests for the GameSim files? Maybe it can be done incrementally by function?

dumbmatter replied:
I don't know... it'd take a lot of tests to catch errors like this (could have happened anywhere), and when bugs like this happen I get a bug report from a user in like 2 minutes lol. So maybe not worth it.
